File: java/270_ClosestBinarySearchTreeValue.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int closestValue(TreeNode root, double target) {
        double dist = Double.MAX_VALUE;
        int closestNode = -1;
        while (root != null) {
            if (abs(target-root.val) < dist) {
                dist = abs(target-root.val);
                closestNode = root.val;
            }
            root = target > root.val ? root.right : root.left;
        }
        return closestNode;
    }
    
    public double abs(double num) {
        return num < 0 ? -num : num;
    }
}

File: swift/270_ClosestBinarySearchTreeValue.swift
class Solution {
    func closestValue(_ root: TreeNode?, _ target: Double) -> Int {
        var node = root
        var closestNode: TreeNode? = node
        var closestDifference = Double(Int.max)
        while node != nil {
            var currentDifference = abs(Double(node!.val) - target)
            if currentDifference < closestDifference {
                closestDifference = currentDifference
                closestNode = node
            }
            node = Double(node!.val) > target ? node?.left : node?.right
        }
        return closestNode!.val
    }
}

class Solution {
    func closestValue(_ root: TreeNode?, _ target: Double) -> Int {
        func traverse(_ node: TreeNode?) {
            guard let node = node else { return }
            
            var currentDifference = abs(Double(node.val) - target)
            if difference > currentDifference {
                closestNode = node
                difference = currentDifference
            }
            traverse(node.left)
            traverse(node.right)
        }
        if root == nil { return 0 }
        var difference = Double(Int.max)
        var closestNode = root
        traverse(root)
        return closestNode?.val ?? 0
    }
}
